optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,416 @@
+                .join(ShiftDate.class,
+                        Joiners.equal((pattern, contract) -> pattern.getFreeDayOfWeek(), ShiftDate::getDayOfWeek))
+                .join(Employee.class,
+                        Joiners.equal((pattern, contractLine, date) -> contractLine.getContract(), Employee::getContract))
+                .ifNotExists(ShiftAssignment.class,
+                        Joiners.equal((pattern, contractLine, date, employee) -> employee, ShiftAssignment::getEmployee),
+                        Joiners.equal((pattern, contractLine, date, employee) -> date.getDayIndex(),
+                                ShiftAssignment::getShiftDateDayIndex))
+                .ifExists(ShiftAssignment.class,
+                        Joiners.equal((pattern, contractLine, date, employee) -> employee, ShiftAssignment::getEmployee),
+                        // TODO: Replace with pair of greaterThan/lessThan joiners?
+                        Joiners.filtering((pattern, contractLine, date, employee,
+                                shift) -> shift.getShiftDateDayIndex() == (date.getDayIndex() + 1)
+                                        || shift.getShiftDateDayIndex() == (date.getDayIndex() + 2)))
+                .penalize("unwantedPatternFreeBefore2DaysWithAWorkDayPattern", HardSoftScore.ONE_SOFT,
+                        (pattern, contractLine, date, employee) -> pattern.getWeight());
+    }
+
+    Constraint unwantedPatternShiftType2DaysPattern(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ShiftType2DaysPattern.class)
+                .join(PatternContractLine.class, Joiners.equal(p -> p, PatternContractLine::getPattern))
+                .join(ShiftAssignment.class,
+                        Joiners.equal((pattern, contractLine) -> pattern.getDayIndex0ShiftType(),
+                                ShiftAssignment::getShiftType),
+                        Joiners.equal((pattern, contractLine) -> contractLine.getContract(), ShiftAssignment::getContract))
+                .join(ShiftAssignment.class,
+                        Joiners.equal((pattern, contractLine, shift) -> shift.getEmployee(), ShiftAssignment::getEmployee),
+                        Joiners.equal((pattern, contractLine, shift) -> shift.getShiftDateDayIndex() + 1,
+                                ShiftAssignment::getShiftDateDayIndex),
+                        Joiners.filtering((pattern, contractLine, shift1, shift2) -> pattern.getDayIndex1ShiftType() == null
+                                || shift2.getShiftType() == pattern.getDayIndex1ShiftType()))
+                .penalize("unwantedPatternShiftType2DaysPattern", HardSoftScore.ONE_SOFT,
+                        (pattern, contractLine, shift1, shift2) -> pattern.getWeight());
+    }
+
+    Constraint unwantedPatternShiftType3DaysPattern(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ShiftType3DaysPattern.class)
```

Comment:
The `PatternContractLine` apparenty includes the `ShiftType3DaysPattern`, therefore we can start from it directly, avoiding the join, and therefore avoiding the `map()` later.